app/controllers/listings_controller.rb

```diff
@@ -371,6 +383,47 @@ def verification_required
 
   private
 
+  def select_unit(params, shape)
+    m_unit = Maybe(shape)[:units].map { |units|
+      shape[:units].length == 1 ? shape[:units].first : parse_unit(params)
```

Comment:
It might make more sense to have unit_belongs_to_shape check here.
Although double tertiary (like the one below) is not readable.
```
shape[:units].length == 1 ? shape[:units].first : unit_belongs_to_shape?(parse_unit(params), shape) ? parse_unit(params) : nil
```